src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,191 @@
+            tail.next = tempNode;
+            tail = tempNode;
+            size++;
+            return;
+        }
+        Node<T> prevNode = head;
+        Node<T> nextNode = head;
+
+        for (int i = 0; i < index; i++) {
+            nextNode = nextNode.next;
+            if (i != index - 1) {
+                prevNode = prevNode.next;
+            }
+        }
+        tempNode.prev = prevNode;
+        prevNode.next = tempNode;
+        tempNode.next = nextNode;
+        nextNode.prev = tempNode;
+
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            Node<T> tempNode = tail;
+            tail = (Node<T>) new Node<>(tempNode, list.get(i), null);
+            tempNode.next = tail;
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (size > index && index >= 0) {
+            Node<T> tempNode = head;
+            for (int i = 0; i < index; i++) {
+                tempNode = tempNode.next;
+            }
+            return tempNode.item;
+        } else {
+            throw new IndexOutOfBoundsException("Invalid Index");
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (size > index && index >= 0) {
+            Node<T> nodeToReplace = head;
+            for (int i = 0; i < index; i++) {
+                nodeToReplace = nodeToReplace.next;
+            }
+            Node<T> newNode = new Node<>(null, value, null);
+            if (index == 0) {
+                newNode.next = head.next;
+                head.next.prev = newNode;
+                head = newNode;
+            } else {
+                newNode.prev = nodeToReplace.prev;
+                newNode.prev.next = newNode;
+                newNode.next = nodeToReplace.next;
+                newNode.next.prev = newNode;
+            }
+            return nodeToReplace.item;
+        } else {
+            throw new IndexOutOfBoundsException("Invalid Index");
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (size > index && index >= 0) {
+            Node<T> nodeToDelete = head;
+            for (int i = 0; i < index; i++) {
+                nodeToDelete = nodeToDelete.next;
+            }
+            if (index == 0 && size == 1) {
+                nodeToDelete.next = null;
+                nodeToDelete.prev = null;
+            } else if (index == 0) {
+                nodeToDelete.next.prev = null;
+                head = nodeToDelete.next;
+            } else if (index == size - 1) {
+                nodeToDelete.prev.next = null;
+                tail = nodeToDelete.prev;
+            } else {
+                nodeToDelete.prev.next = nodeToDelete.next;
+                nodeToDelete.next.prev = nodeToDelete.prev;
+            }
+            size--;
+            return nodeToDelete.item;
```

Comment:
too complex logic, let's first find a node that will be removed, after that I guess you could see the ways of simplifying of your logic